Adding Snapshot Taking document to public (#1)
LiskHQ/lisk-token-claim: README.md

# lisk-token-claim

This library is the monorepo for:

### Tree Builder

Builds Merkle Tree from a snapshot and computes Merkle Root.

## Setup and Installation

```

> $ git clone git@github.com:LiskHQ/lisk-token-claim.git && cd lisk-token-claim

Author: probably clone part is not so important in readme. everyone should know how to clone

Reviewer: @AndreasKendziorra sorry I'm removing it😿

# Lisk Token Claim

This library builds Merkle Tree and build Merkle Root

## Installation

```

> $ yarn

Reviewer 2: I guess you first need to clone and then go into the directory, right?

Reviewer 2: I was actually serious that this should be added to the README :) but ignore it if you don't agree

Reviewer: Added :D

```

## Run

```
$ yarn generate-merkle-tree             # Using ./data/mainnet
$ yarn generate-merkle-tree-testing     # Using ./data/example
```

## Files

| Name                                      | Description                                         | Generated By                  |
| ----------------------------------------- | --------------------------------------------------- | ----------------------------- |
| `/data/<network>/balance.json`            | Stores balance by address after a snapshot is taken | Snapshot                      |
| `/data/<network>/merkle-tree-result.json` | Stores MerkleRoot, and node for each account        | `$ yarn generate-merkle-tree` |

## Merkle Node
Each node will be encoded as:
```
<LSK_ADDRESS_IN_HEX><BALANCE_IN_BEDDOWS><NUMBER_OF_SIGNATURES><MANDATORY_KEYS><OPTIONAL_KEYS>
P.S. If the address is not a multisig address, NUMBER_OF_SIGNATURES would be 0
```

## Params

```
balance.json:
{
    lskAddress: string;
    balance: number;
    numberOfSignatures?: number;
    mandatoryKeys?: string[];
    optionalKeys?: string[];
}

merkle-tree-result.json:
{
    merkleRoot: string;
    nodes: {
        lskAddress: string;

>         address: string;

Reviewer 2: what is this needed for?

Reviewer: This is for the ease of cross-checking with the json/csv generated from the snapshot manually.

        balance: number;

>         balanceBeddows: number;

Reviewer 2: why the two different values?

Reviewer: Same as above, for manually checking. The tree builder will use beddows instead.

# Lisk Token Claim

> This library builds Merkle Tree and build Merkle Root

Reviewer 3: ```suggestion
This library builds Merkle Tree from a snapshot and computes Merkle Root.
```

Reviewer: Updated

## Installation

```
$ yarn
```

## Run

```
$ yarn generate-merkle-tree             # Using ./data/mainnet
$ yarn generate-merkle-tree-testing     # Using ./data/example
```

## Files

| Name                                      | Description                                         | Generated By                  |
| ----------------------------------------- | --------------------------------------------------- | ----------------------------- |
| `/data/<network>/balance.json`            | Stores balance by address after a snapshot is taken | Snapshot                      |
| `/data/<network>/merkle-tree-result.json` | Stores MerkleRoot, and node for each account        | `$ yarn generate-merkle-tree` |

## Merkle Node

Each node will be encoded as:

```
<LSK_ADDRESS_IN_HEX><BALANCE_IN_BEDDOWS><NUMBER_OF_SIGNATURES><MANDATORY_KEYS><OPTIONAL_KEYS>
P.S. If the address is not a multisig address, NUMBER_OF_SIGNATURES would be 0
```

## Params

```
balance.json:
{
    lskAddress: string;

>     balance: number;

Reviewer 3: `balanceBeddows` is missing?

Reviewer: Updated

    numberOfSignatures?: number;
    mandatoryKeys?: string[];
    optionalKeys?: string[];
}

merkle-tree-result.json:
{
    merkleRoot: string;

>     nodes: {

Reviewer 3: `leaves` probably?

Reviewer: Updated

        lskAddress: string;
        address: string;
        balance: number;
        balanceBeddows: number;
        numberOfSignatures: number;
        mandatoryKeys: string[];
        optionalKeys: string[];

>         payload: string;

Reviewer 3: Is `payload` used somewhere?
In `merkle-tree-result.json` file I can't find it.

Reviewer: payload is a field to manually compare with existing data, however OZ's lib does not have a function to generate abi-encoded message so now I have removed it

        hash: string;
        proof: string[];
    }[];
}
```

> 

Reviewer 3: Why is structure of `signatures.json` not listed here?

Reviewer: Updated

## _Testing Only_

### Generate balance.json from `dev-validators.json`

```
$ yarn create-balances-testing
```

### (After running `generate-merkle-tree`) Sign all accounts

```
$ yarn sign-testing
```

### `merkle-tree-simple`

Due to the constraint of vm.parseJson in Foundry, fields in `merkle-tree-result` such as decimal field will trigger error.

> `merkle-tree-simple.json` is a lightweight version of `merkle-tree.json`, which also has sufficient params for claiming.

Reviewer 3: Are files `merkle-tree-simple.json` and `merkle-tree.json` generated by some script/external SW or are they missing in this commit?

Reviewer: Updated

> This library builds Merkle Tree from a snapshot and computes Merkle Root.

Reviewer 2: As the deployment script for the claims process uses now a json file that contains only the Merkle root, it would probably be helpful to additionally generate this file with the scripts here as well. Do you agree?
If so, I would call it (and rename the one in the deployment script to) `merkle-root.json` or similar instead of `merkle-tree.json`.

Reviewer: Can do

## Setup and Installation

```
$ git clone git@github.com:LiskHQ/lisk-token-claim.git && cd lisk-token-claim
$ yarn
```

## Run

```
$ yarn generate-merkle-tree:mainnet     # Using ./data/mainnet
$ yarn generate-merkle-tree:testnet     # Using ./data/testnet
$ yarn generate-merkle-tree:example     # Using ./data/example
```

## Files

| Name                                             | Description                                                                                                                                 | Generated By                            |
| ------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------- |-----------------------------------------|

> | `/data/<network>/accounts.json`                  | Stores addresses, balances, and multisig details(If any) by address after a snapshot is taken, addresses must be sorted in ascending order. | Snapshot                                |

Reviewer 2: ```suggestion
| `/data/<network>/accounts.json`                  | Stores addresses, balances, and multisig details (If any) per account after a snapshot is taken, addresses must be sorted in ascending order. | Snapshot                                |
```

Reviewer: Updated

> | `/data/<network>/accounts.json`                  | Stores addresses, balances, and multisig details(If any) by address after a snapshot is taken, addresses must be sorted in ascending order. | Snapshot                                |

Reviewer 2: An additional hint that this is used as the input for Merkle tree computation would be helpful. Once we know how the snapshot and this file are computed, it would additionally be good to mention this here as well.

Reviewer: Updated

```
$ yarn generate-merkle-tree             # Using ./data/mainnet
$ yarn generate-merkle-tree:example     # Using ./data/example
```

## Files

| Name                                             | Description                                                                                       | Generated By                  |
|--------------------------------------------------|---------------------------------------------------------------------------------------------------| ----------------------------- |

> | `/data/<network>/balance.json`                   | Stores balance by address after a snapshot is taken, addresses must be sorted in ascending order. | Snapshot                      |

Reviewer 2: it also stores information about keys

## Files

| Name                                               | Description                                                                                                                                                                           | Generated By                            |
|----------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| --------------------------------------- |
| `/data/<network>/accounts.json`                    | Stores addresses, balances, and multisig details(If any) per account after a snapshot is taken, addresses must be sorted in ascending order. Will be used for MerkleTree computation. | Snapshot                                |
| `/data/<network>/merkle-tree-result-detailed.json` | Stores MerkleRoot, and leaves for each account. Will be used for examination by 3rd Party or public.                                                                                  | `$ yarn generate-merkle-tree:<network>` |

> | `/data/<network>/merkle-tree-result.json`          | Stores MerkleRoot, and leaves for each account. Will be used for testing and simulation for Claim Contract.                                                                           | `$ yarn generate-merkle-tree:<network>` |

Reviewer 2: I'm not 100%ly sure what you mean with "simulation for Claim Contract". Will it be used by the UI to get the inclusion proof?

merkle-tree-result.json:
{
    merkleRoot: string;
    leaves: {
        lskAddress: string;
        address: string;
        balance: number;
        balanceBeddows: number;
        numberOfSignatures: number;
        mandatoryKeys: string[];
        optionalKeys: string[];
        hash: string;
        proof: string[];
    }[];
}

merkle-tree-result-simple.json:
{
  merkleRoot: string;
  leaves: {
    b32Address: string;
    balanceBeddows: number;
    mandatoryKeys: string[];
    numberOfSignatures: number;
    optionalKeys: string[];
    proof: string[];
  }[];
}
# `address` is a reserved in solidity, hence `b32Address` here
```

> ### `merkle-tree-simple`

Reviewer 2: Is there really a need to create two merkle tree json files? wouldn't the regular one be sufficient?

Reviewer 2: If the Foundry limitation with regard to decimal numbers is the reason to the create the "simple" version of the json file, I would rather remove the decimals number from the json, i.e. use only integers to express balances in beddows.

Reviewer: The philosophy of having 2 files is to make `merkle-tree-result` as detailed as possible, even if some fields are very trivial. As that file would be published to and reviewed by third party, by do that those params could be even be verified by naked eye.

Reviewer 2: 1. Is my understanding correct that `merkle-tree-result-simple.json` will be used by the UI (or its backend service) to create inclusion proofs, and `merkle-tree-result.json` is only intended to be read by humans? If so, I would say that the first one is THE merkle tree file and I would not add the term "simple" to its name. So maybe
`merkle-tree-result-simple.json` --> `merkle-tree-result.json`
`merkle-tree-result.json` --> `merkle-tree-result-detailed.json`
2. It would be good to extend the description in the [Files](https://github.com/LiskHQ/lisk-token-claim/blob/a9037d697a9043f271f82066f41520faca73ee6c/README.md#files) section such that it's clear for everyone what's the purpose of the individual files.

Reviewer: `merkle-tree-result.json` will be read by 3rd party, and the public, they would do it however that like.

But I agree renaming it to `detailed`

merkle-tree-result.json:
{
    merkleRoot: string;
    leaves: {
        lskAddress: string;
        address: string;
        balance: number;
        balanceBeddows: number;
        numberOfSignatures: number;
        mandatoryKeys: string[];
        optionalKeys: string[];
        hash: string;
        proof: string[];
    }[];
}

merkle-tree-result-simple.json:
{
  merkleRoot: string;
  leaves: {
    b32Address: string;
    balanceBeddows: number;
    mandatoryKeys: string[];
    numberOfSignatures: number;
    optionalKeys: string[];
    proof: string[];
  }[];
}
# `address` is a reserved in solidity, hence `b32Address` here
```

### `merkle-tree-simple`

Due to the constraint of vm.parseJson in Foundry, fields in `merkle-tree-result` such as decimal field will trigger error.

> `merkle-tree-simple.json` is a lightweight version of `merkle-tree-result.json`, which also has sufficient params for claiming.

Reviewer 3: `merkle-tree-simple.json` -> `merkle-tree-result-simple.json`?

> 

Reviewer 2: the heading in line 86 must also be updated

> Due to the constraint of vm.parseJson in Foundry, fields in `merkle-tree-result` such as decimal field will trigger error.

Reviewer 2: ```suggestion
Due to the constraint of vm.parseJson in Foundry, fields in `merkle-tree-result-detailed` such as decimal field will trigger error.
```

Reviewer: This paragraph is removed

> `merkle-tree-simple.json` is a lightweight version of `merkle-tree-result-detailed.json`, which also has sufficient params for claiming.

Reviewer 2: ```suggestion
`merkle-tree-result.json` is a lightweight version of `merkle-tree-result-detailed.json`, which also has sufficient params for claiming.
```

Reviewer: This paragraph is removed
